contrib/platform/src/com/sun/jna/platform/win32/Cfgmgr32Util.java

```diff
@@ -96,4 +95,160 @@ public static String CM_Get_Device_ID(int devInst) throws Cfgmgr32Exception {
             return buffer.getWideString(0);
         }
     }
+
+    /**
+     * Utility method to call Cfgmgr32's CM_Get_DevNode_Registry_Property that
+     * allocates the required memory for the Buffer parameter, and provides getters
+     * to retrieve values of the appropriate type.
+     *
+     * @param devInst
+     *            Caller-supplied device instance handle that is bound to the local
+     *            machine.
+     * @param ulProperty
+     *            A {@code CM_DRP_}-prefixed constant value that identifies the
+     *            device property to be obtained from the registry. These constants
+     *            are defined in Cfgmgr32.h.
+     * @return A {@link RegistryProperty} object encapsulating information about the
+     *         specified registry property for the device. If no value exists for
+     *         this property, returns null
+     * @throws Cfgmgr32Exception
+     */
+    public static RegistryProperty CM_Get_DevNode_Registry_Property(int devInst, int ulProperty)
+            throws Cfgmgr32Exception {
+
+        // Get byte count and type
+        IntByReference size = new IntByReference();
+        IntByReference type = new IntByReference();
+        int ret = Cfgmgr32.INSTANCE.CM_Get_DevNode_Registry_Property(devInst, ulProperty, type, null, size, 0);
+        // This should fail with CR_BUFFER_SMALL
+        // Other errors such as CR_NO_SUCH_VALUE should return null
+        if (ret != Cfgmgr32.CR_BUFFER_SMALL) {
+            return null;
+        }
+        // It is possible to have a valid value with type but 0 size
+        if (size.getValue() == 0) {
+            return new RegistryProperty(null, type.getValue());
+        }
+        Memory buffer = new Memory(size.getValue());
+
+        ret = Cfgmgr32.INSTANCE.CM_Get_DevNode_Registry_Property(devInst, ulProperty, type, buffer, size, 0);
+        if (ret != Cfgmgr32.CR_SUCCESS) {
+            throw new Cfgmgr32Exception(ret);
+        }
+        return new RegistryProperty(buffer, type.getValue());
+    }
+
+    /**
+     * A class encapsulating the buffer returned from
+     * {@link Cfgmgr32#CM_Get_DevNode_Registry_Property()} with type safe getters
+     * for convenience.
+     */
+    public static class RegistryProperty {
```

Comment:
Serious question: what does `RegistryProperty` give us, that `java.lang.Object` does not provide? Getting rid of native memory early would have the benefit of getting native memory pressure down (if you fetch _many_ registry values, all of these are held in tiny Memory object, that potentially hold much more data off-heap).

Reply:
Well, the serious answer is "You don't need to cast."  But I see your point that they are functionally equivalent.    If you know  you need `foo.stringValue()` in my solution, you equally know you need to do `(String) foo` in yours.

Reply:
So I finally came around to the realization of the simplicity and elegancy of your original suggestion.